### ISSUE TITLE
Introduce a selection mask

### ISSUE THREAD

**Issue by robert3005** [vortex-data/vortex#1270]
When performing filters it can be faster to push the mask from previous filter and apply the filter than apply the filter and combine the results

related #1180

**Comment by danking** [comment]
Proposal: always apply any RowMask (whether from indices or previous filters) to the _validity_ of the array. This moves the responsibility for deciding between take-then-filter and filter-then-take to each encoding's filter function (where it benefits any user of filter!).

This also means that boolean array encodings are responsible for a fast implementation of `to_bitmap`, which, again, benefits anyone working with boolean arrays rather than just the layout subsystem.

cc: @gatesn

**Comment by gatesn** [comment]
So an idea for this is something like:

```
trait CompareFn {
   fn compare(&self, lhs: A, rhs: A) -> VortexResult<...>;

   fn compare_with_mask(&self, lhs: A, rhs: A, mask: &FilterMask) -> VortexResult<...> {
       self.compare(lhs, rhs)
   }
}
```

Where the `with_mask` variant of the function returns a result with the same length as the input array (NOT the true_count of the mask). It guarantees that results in the true positions of the mask are correct, but makes no guarantee about the false positions.

**Comment by gatesn** [comment]
Our latest plan here is to look at adding a `vortex_mask::Mask` to every ArrayData representing a selection mask. This allows us to perform partial computation and deferred filtering.

**Comment by gatesn** [comment]
For reference, the "Selection Pushdown" paper https://www.microsoft.com/en-us/research/uploads/prod/2023/06/parquet-select-sigmod23.pdf

**Comment by danking** [comment]
I've been mulling over this issue a bit and reading up on the references. Here's some notes I have so far.

---

The arrow-rs repository has an issue requesting an "evaluate with selection mask": https://github.com/apache/arrow-rs/issues/3620. Their conclusion seemed to be that a selection vector does not clearly provide more functionality than invalidating those values, computing, and filtering to the desired values.

Andy Grove proposed a similar API to datafusion but ultimately [decided it wasn't worth pursuing](https://github.com/apache/datafusion/issues/5944#issuecomment-2260794953). It seems that he found that construction of a list of indices was too expensive in most cases compared to filtering.

Li, et al. "Selection Pushdown ..." describes their key finding as:
> A key finding in this paper is the recognition that the Bit Manipulation Instructions (BMI), which were introduced by Intel in 2013 and are now available in nearly all Intel and AMD processors, are well-suited to address thischallenge. To the best of our knowledge, this is the first paper that leverages BMI for database applications. It is important to note that applying BMI to this application is _nontrivial_, which likely explains why BMI has been largely ignored in database systems over the past decade. Additionally, we find that the role of BMI goes well beyond the select operator. In fact, BMI plays a crucial role in all the key techniques proposed in this paper, as summarized below

In describing their "select operator" they say:
> An obvious solution to this problem would be to scan over all bit-packed values, extracting and gathering selected bit-packed values one at a time, which runs in 𝑂 (𝑛) instructions. However, considering that each value is typically only a few bits long and much smaller than the processor word (e.g., 64 bits), this simple implementation does not fully utilize the width of a processor word, thus wasting the parallelism available in processors.
>
> Hence, our goal is to design a bit-parallel select operator. Intuitively, this means that the algorithm is able to simultaneously process all values that are packed into a processor word, moving all selected values to appropriate positions in parallel.

I haven't finished the paper yet, so take this with a grain of salt, but: Li, et al. seem to focus on the bitpacked codes of a dictionary encoded column:
> The most widely-usedencoding scheme is dictionary encoding, where each distinct value in a column is mapped to a unique small code according to a dictionary built for the column. These small codes of column values are usually stored in a bit-packed manner, using as many bits as needed to represent each code. This encoding scheme, though effective in reducing storage cost, imposes challenges on query processing as these codes must be decoded before they can be further processed.

I'll finish the paper, but my first impression is that the bigger insight for us is how to improve BitPacked Encoding's take.

**Comment by danking** [comment]
My initial assessment seems to be correct. Li, et al. is primarily about the value of PEXT (parallel bit extract) and PDEP (parallel bit deposit). These operations are available on both Intel and AMD x86 chips. In ARM, these operations are called BEXT and BDEP. They're part of Scalable Vector Extensions 2 (SVE2), which is available in ARMv9.0-A. Apple M3 uses ARMv8.6; however the new Apple M4s support ARMv9!

AFAICT, Rust has not added BEXP or BDEP to core::arch::aarch64 yet. Their x86_64 equivalents do exist. Arrow-cpp does use [pext](https://github.com/apache/arrow/blob/main/cpp/src/parquet/level_conversion_inc.h#L269) when available to manipulate definition levels. It's possible that using them to implement bit unpack in fastlanes would produce a noticeable improvement on the benchmark machines.

Overall I think we should close this issue. @a10y and @joseph-isaacs also did a pretty extensive evaluation of selection masks in practice in Vortex and did not find a significant improvement from them.

---

Based on Li, et al.'s benchmarks, it appears that dictionary and run-length decoding is a substantial fraction of TPC-H runtime. Figure 14 shows the speed up in a "standalone" implementation of the SQL queries. Notice the y-axis scales vary significantly.

<img width="969" alt="Image" src="https://github.com/user-attachments/assets/6fd8b3b1-7b17-416d-b46b-60a3d2effb25" />

Figure 16 is a bit more "real" showing the speed up in Spark. Still meaningful improvements.

<img width="644" alt="Image" src="https://github.com/user-attachments/assets/ac303567-4d5a-4dfa-8af1-34b17f015ea0" />

They note that the speed up is well correlated with the selectivity of the filter on the Lineitem table.
<img width="642" alt="Image" src="https://github.com/user-attachments/assets/31785bd7-fe92-4f5e-bbc9-cbbc46c11a4a" />

**Comment by gatesn** [comment]
We've decided to not do this for now, but with the new #2463, there's nothing to stop individual arrays from holding selection masks if they wish
